### PR TITLE
Fix #7. Introduce tags

### LINKED ISSUE #7: New syntax for meta-data
#### Goal
Provide a simple means for defining private meta-data for messages.

#### Description
Currently, meta-data can be added to messages by using traits. Traits without namespaces are considered private.

```
brand-name =Firefox
  [gender] masculine
```

#5 and #6 will simplify traits and we'll need a new way to encode meta-data.

The proposal is to use binary tags attached to the value:

```
#masculine
brand-name = Firefox
```

The benefit of the binary approach is that there's usually no need to name the property in question (_gender_).

#### Discussion
https://groups.google.com/forum/#!topic/mozilla.tools.l10n/dhWfBXHzuZI

### PR DESCRIPTION
**Pull request by stasm**
Tags are binary values attached to messages.  They are language-specific and
can be used to describe grammatical characteristics of the message.

    brand-name = Firefox
        #masculine

    brand-name = Aurora
        #feminine
        #vowel

Tags can be used in select expressions by matching a hashtag name to the
message:

    has-updated = { brand-name ->
        [masculine] …
        [feminine] …
       *[other] …
    }

Tags can only be defined on messages which have a value and don't have any
attributes.

This PR depends on #32.

### CODE REVIEW

File: syntax/fluent.ebnf [hunk 1]
```diff
@@ -1,41 +1,53 @@
...
+
+/* special chars: { } [ \ ] */
+special              ::= [#x5b#x5c#x5d#x7b#x7d]
+
+/* line feed, carriage return; space, tab */
+line-break           ::= [#xA#xD]+
+inline-space         ::= [#x20#x9]+
+
+/* if you break lines you need to indent afterwards */
+break-indent         ::= (line-break* inline-space)+
```
Comment (zbraniecki) [inline syntax/fluent.ebnf]: This seems to forbid the edge case where you do "\n\n  Indented test" right?

File: syntax/fluent.ebnf [hunk 2]
```diff
@@ -1,41 +1,53 @@
...
+
+_                    ::= inline-space
+NL                   ::= line-break
+__                   ::= break-indent
```
Comment (zbraniecki) [inline syntax/fluent.ebnf]: what's the reason behind defining two keys for `inline-space` and two for `break-indent`?

File: syntax/fluent.ebnf [hunk 3]
```diff
@@ -1,41 +1,53 @@
...
-unquoted-pattern     ::= (unquoted-text | placeable | block-text)+
-quoted-pattern       ::= '"' (quoted-text | placeable)* '"'
-unquoted-text        ::= ([^{] | '\{')+
-quoted-text          ::= ([^{"] | '\{' | '\"')+
-block-text           ::= NL __ '|' unquoted-pattern
-
-placeable            ::= '{' __ (expression | select-expression | variant-list) __ '}'
-expression           ::= quoted-pattern
+variant-symbol       ::= word (_ word)*
+variant              ::= NL __ '[' _ variant-key _ ']' __ pattern
```
Comment (zbraniecki) [inline syntax/fluent.ebnf]: so, variants are required not to have white spaces?
This will make 
```
key = { $foo ->
  *[one] One
}
```

not work

File: syntax/fluent.ebnf [hunk 4]
```diff
@@ -1,41 +1,53 @@
...
+default-variant      ::= NL __ '*[' _ variant-key _ ']' __ pattern
+variant-list         ::= variant* default-variant variant*
+
+tag                  ::= '#' word
```
Comment (zbraniecki) [inline syntax/fluent.ebnf]: I'm a bit concerned about the ability to put any character here. The idea that:
```
key = Value
  #####
  #!
  #@
  #⁂
  #🍆
```

will  work and is a proper tag is a bit scary and feels error prone and may make it harder to recover.

File: syntax/fluent.ebnf [hunk 5]
```diff
@@ -1,41 +1,53 @@
...
+tag-list             ::= NL (__ tag)+
+
+attribute            ::= NL __ '.' identifier value
+attribute-list       ::= attribute+
+
+message              ::= identifier ((value tag-list?) | (value? attribute-list))
+value                ::= __ '=' __ pattern
```
Comment (zbraniecki) [inline syntax/fluent.ebnf]: so, this opens up
```
key
  = Value
```

I have to say I don't like it from the readability perspective :(

Comment (zbraniecki) [inline syntax/fluent.ebnf]: It also forbids `key=Value` format (enforcing whitespace).

File: syntax/fluent.ebnf (hunk 1 above)
Comment (stasm) [inline syntax/fluent.ebnf]: Sorry, I don't understand this. Can you provide a full example? Feel free to use `_` to denote spaces for better visibility.

File: syntax/fluent.ebnf (hunk 2 above)
Comment (stasm) [inline syntax/fluent.ebnf]: Are you asking about why `inline-space` and `break-indent` have aliases? If so, the answer is simply: readability of other productions.

File: syntax/fluent.ebnf (hunk 3 above)
Comment (stasm) [inline syntax/fluent.ebnf]: I don't understand this either. Do you mean variant keys or their patterns?

File: syntax/fluent.ebnf (hunk 4 above)
Comment (stasm) [inline syntax/fluent.ebnf]: I think we should add `#` to the list of special characters, just in case. I don't see how `🍆` is different from `żółć`, though :)

File: syntax/fluent.ebnf (hunk 5 above)
Comment (stasm) [inline syntax/fluent.ebnf]: I think @Pike wasn't convinced about the line-break before `=` either, so I'll remove it. Especially that we enforce the (pattern, tags, attrs) order anyways.

> It also forbids key=Value format (enforcing whitespace).

Would you like me to add it back?

File: syntax/fluent.ebnf (hunk 1 above)
Comment (zbraniecki) [inline syntax/fluent.ebnf]: oh, my bad, misread the syntax. Sorry!

File: syntax/fluent.ebnf (hunk 3 above)
Comment (zbraniecki) [inline syntax/fluent.ebnf]: That this enforces
```
key = { $foo ->
  *[ one ] One
}
```
whitespaces within `[`, `]`.

File: syntax/fluent.ebnf (hunk 4 above)
Comment (zbraniecki) [inline syntax/fluent.ebnf]: yeah, I just feel like i'd prefer to stay on the side of [a-z] initially and see if anyone asks for more, but maybe it'll work just fine. :)

File: syntax/fluent.ebnf (hunk 5 above)
Comment (zbraniecki) [inline syntax/fluent.ebnf]: I don't see a reason to forbid it.

File: syntax/fluent.ebnf (hunk 3 above)
Comment (stasm) [inline syntax/fluent.ebnf]: Duh, of course! Thanks, great catch :) I'll change to `_?` where needed.

File: syntax/fluent.ebnf (hunk 4 above)
Comment (stasm) [inline syntax/fluent.ebnf]: Just for Polish, I'd like to make `#męski` and `#żeński` possible, and I expect other localizers to want to use tags in their own languages, too.  This is also good for documentation and examples, as it clearly demonstrates that tags are language-specific.
